### ISSUE TITLE
pseudo Dice: 0.0

### ISSUE THREAD

**Issue by ltt648** [MIC-DKFZ/nnUNet#2623]
Hello, I encountered the following problem while using the MSD Test10 colon dataset to segment lesions. 
2024-11-27 15:39:30.635207: Epoch 0 
2024-11-27 15:39:30.635375: Current learning rate: 0.01 
2024-11-27 15:45:23.622632: train_loss 0.0728 
2024-11-27 15:45:23.622912: val_loss -0.0007 
2024-11-27 15:45:23.622983: Pseudo dice [0.0] 
2024-11-27 15:45:23.623044: Epoch time: 352.99 s 
2024-11-27 15:45:23.623091: Yayy! New best EMA pseudo Dice: 0.0 
pseudo Dice is 0 in first Epoch,although it increase to 70 after 1000 Epochs, it is still very low,how can i do for that?

**Comment by jmayank23** [comment]
I am facing a similar problem, at epoch 300 the pseudo dice is 0. 

For reference, here is the dataset.json file
`{
    "channel_names": {
        "0": "grayscale"
    },
    "labels": {
        "background": 0,
        "vessel": 1
    },
    "numTraining": 61068,
    "file_ending": ".png"
}`

and here is a snippet of the training output

############################
INFO: You are using the old nnU-Net default plans. We have updated our recommendations. Please consider using those instead! Read more here: https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/resenc_presets.md
############################

Using device: cuda:0
/home/mayank/anaconda3/envs/cta_pad_py310/lib/python3.10/site-packages/nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py:164: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  self.grad_scaler = GradScaler() if self.device.type == 'cuda' else None

#######################################################################
Please cite the following paper when using nnU-Net:
Isensee, F., Jaeger, P. F., Kohl, S. A., Petersen, J., & Maier-Hein, K. H. (2021). nnU-Net: a self-configuring method for deep learning-based biomedical image segmentation. Nature methods, 18(2), 203-211.
#######################################################################

2025-01-10 03:15:57.907878: do_dummy_2d_data_aug: False
2025-01-10 03:15:58.112132: Using splits from existing split file: /home/mayank/Projects/CTA_Data/nnunet/preprocessed/Dataset123_VesselSeg/splits_final.json
2025-01-10 03:15:58.136438: The split file contains 5 splits.
2025-01-10 03:15:58.136529: Desired fold for training: 0
2025-01-10 03:15:58.136562: This split has 48854 training and 12214 validation cases.
using pin_memory on device 0
using pin_memory on device 0
/home/mayank/anaconda3/envs/cta_pad_py310/lib/python3.10/site-packages/torch/optim/lr_scheduler.py:62: UserWarning: The verbose parameter is deprecated. Please use get_last_lr() to access the learning rate.
  warnings.warn(

This is the configuration used by this training:
Configuration name: 2d
 {'data_identifier': 'nnUNetPlans_2d', 'preprocessor_name': 'DefaultPreprocessor', 'batch_size': 57, 'patch_size': [224, 256], 'median_image_size_in_voxels': [217.0, 256.0], 'spacing': [1.0, 1.0], 'normalization_schemes': ['ZScoreNormalization'], 'use_mask_for_norm': [False], 'resampling_fn_data': 'resample_data_or_seg_to_shape', 'resampling_fn_seg': 'resample_data_or_seg_to_shape', 'resampling_fn_data_kwargs': {'is_seg': False, 'order': 3, 'order_z': 0, 'force_separate_z': None}, 'resampling_fn_seg_kwargs': {'is_seg': True, 'order': 1, 'order_z': 0, 'force_separate_z': None}, 'resampling_fn_probabilities': 'resample_data_or_seg_to_shape', 'resampling_fn_probabilities_kwargs': {'is_seg': False, 'order': 1, 'order_z': 0, 'force_separate_z': None}, 'architecture': {'network_class_name': 'dynamic_network_architectures.architectures.unet.PlainConvUNet', 'arch_kwargs': {'n_stages': 6, 'features_per_stage': [32, 64, 128, 256, 512, 512], 'conv_op': 'torch.nn.modules.conv.Conv2d', 'kernel_sizes': [[3, 3], [3, 3], [3, 3], [3, 3], [3, 3], [3, 3]], 'strides': [[1, 1], [2, 2], [2, 2], [2, 2], [2, 2], [2, 2]], 'n_conv_per_stage': [2, 2, 2, 2, 2, 2], 'n_conv_per_stage_decoder': [2, 2, 2, 2, 2], 'conv_bias': True, 'norm_op': 'torch.nn.modules.instancenorm.InstanceNorm2d', 'norm_op_kwargs': {'eps': 1e-05, 'affine': True}, 'dropout_op': None, 'dropout_op_kwargs': None, 'nonlin': 'torch.nn.LeakyReLU', 'nonlin_kwargs': {'inplace': True}, 'deep_supervision': True}, '_kw_requires_import': ['conv_op', 'norm_op', 'dropout_op', 'nonlin']}, 'batch_dice': True} 

These are the global plan.json settings:
 {'dataset_name': 'Dataset123_VesselSeg', 'plans_name': 'nnUNetPlans', 'original_median_spacing_after_transp': [999.0, 1.0, 1.0], 'original_median_shape_after_transp': [1, 217, 256], 'image_reader_writer': 'NaturalImage2DIO', 'transpose_forward': [0, 1, 2], 'transpose_backward': [0, 1, 2], 'experiment_planner_used': 'ExperimentPlanner', 'label_manager': 'LabelManager', 'foreground_intensity_properties_per_channel': {'0': {'max': 255.0, 'mean': 247.1350673568466, 'median': 255.0, 'min': 0.0, 'percentile_00_5': 157.0, 'percentile_99_5': 255.0, 'std': 15.43940059304963}}} 

2025-01-10 03:16:03.730677: unpacking dataset...
2025-01-10 03:16:09.048312: unpacking done...
2025-01-10 03:16:09.081988: Unable to plot network architecture:
2025-01-10 03:16:09.082060: No module named 'hiddenlayer'
2025-01-10 03:16:09.086818: 
2025-01-10 03:16:09.086955: Epoch 0
2025-01-10 03:16:09.087069: Current learning rate: 0.01
2025-01-10 03:16:41.121617: train_loss 0.017
2025-01-10 03:16:41.121762: val_loss -0.0228
2025-01-10 03:16:41.121819: Pseudo dice [np.float32(0.0)]
2025-01-10 03:16:41.121878: Epoch time: 32.04 s
2025-01-10 03:16:41.121923: Yayy! New best EMA pseudo Dice: 0.0
2025-01-10 03:16:42.050638: 
2025-01-10 03:16:42.050786: Epoch 1
2025-01-10 03:16:42.050869: Current learning rate: 0.00999
2025-01-10 03:17:12.548879: train_loss -0.0451
2025-01-10 03:17:12.549037: val_loss -0.0565
2025-01-10 03:17:12.549104: Pseudo dice [np.float32(0.0)]
2025-01-10 03:17:12.549162: Epoch time: 30.5 s
2025-01-10 03:17:13.212064: 
2025-01-10 03:17:13.212212: Epoch 2
2025-01-10 03:17:13.212297: Current learning rate: 0.00998
2025-01-10 03:17:44.340477: train_loss -0.0621
2025-01-10 03:17:44.340607: val_loss -0.0655
2025-01-10 03:17:44.340709: Pseudo dice [np.float32(0.0)]
2025-01-10 03:17:44.340776: Epoch time: 31.13 s
2025-01-10 03:17:45.024884: 
2025-01-10 03:17:45.025084: Epoch 3
2025-01-10 03:17:45.025168: Current learning rate: 0.00997
2025-01-10 03:18:16.356830: train_loss -0.0629
2025-01-10 03:18:16.356983: val_loss -0.0787
2025-01-10 03:18:16.357041: Pseudo dice [np.float32(0.0004)]
2025-01-10 03:18:16.357102: Epoch time: 31.33 s
2025-01-10 03:18:16.357150: Yayy! New best EMA pseudo Dice: 0.0
2025-01-10 03:18:17.901233: 
2025-01-10 03:18:17.901428: Epoch 4
2025-01-10 03:18:17.901494: Current learning rate: 0.00996
2025-01-10 03:18:49.158918: train_loss -0.0789
2025-01-10 03:18:49.159060: val_loss -0.0942
2025-01-10 03:18:49.159122: Pseudo dice [np.float32(0.05)]
2025-01-10 03:18:49.159184: Epoch time: 31.26 s
2025-01-10 03:18:49.159235: Yayy! New best EMA pseudo Dice: 0.004999999888241291
.
.
.
2025-01-10 05:52:54.651281: Epoch 297
2025-01-10 05:52:54.651384: Current learning rate: 0.00728
2025-01-10 05:53:25.754518: train_loss -0.215
2025-01-10 05:53:25.754981: val_loss -0.2384
2025-01-10 05:53:25.755073: Pseudo dice [np.float32(0.0)]
2025-01-10 05:53:25.755151: Epoch time: 31.1 s
2025-01-10 05:53:26.475422: 
2025-01-10 05:53:26.475611: Epoch 298
2025-01-10 05:53:26.475668: Current learning rate: 0.00727
2025-01-10 05:53:57.588746: train_loss -0.1961
2025-01-10 05:53:57.588876: val_loss -0.2185
2025-01-10 05:53:57.588930: Pseudo dice [np.float32(0.0)]
2025-01-10 05:53:57.588987: Epoch time: 31.11 s
2025-01-10 05:53:58.303989: 
2025-01-10 05:53:58.304088: Epoch 299
2025-01-10 05:53:58.304160: Current learning rate: 0.00726
2025-01-10 05:54:29.430572: train_loss -0.2073
2025-01-10 05:54:29.430705: val_loss -0.1502
2025-01-10 05:54:29.430760: Pseudo dice [np.float32(0.0)]
2025-01-10 05:54:29.430818: Epoch time: 31.13 s
2025-01-10 05:54:30.606046: 
2025-01-10 05:54:30.606178: Epoch 300
2025-01-10 05:54:30.606261: Current learning rate: 0.00725
2025-01-10 05:55:01.604056: train_loss -0.1955
2025-01-10 05:55:01.604242: val_loss -0.2562
2025-01-10 05:55:01.604297: Pseudo dice [np.float32(0.0)]
2025-01-10 05:55:01.604375: Epoch time: 31.0 s

**Comment by Lars-Kraemer** [comment]
Hey @ltt648 

A pseudo Dice of 0 in the first epoch is not unusual, especially for difficult small-target tasks like MSD Colon. Pseudo Dice in nnU-Net is computed on validation patches during training and is only a rough sanity-check metric; it is not the same as the final Dice on full validation cases. Also, colon cancer segmentation is one of the hardest MSD tasks, with much lower Dice scores than easier tasks in the challenge summary. If your final full-case Dice is around 0.7, I guess that is not bad for this task

Best,
Lars

**Comment by Lars-Kraemer** [comment]
Hey @jmayank23 

Also here pseudo Dice of 0 for the first few epochs can be normal, but if it is still 0 at epoch 300, I would first suspect a dataset issue. Please verify that your PNG labels are true integer class maps with values 0/1 (not 0/255, RGB, or anti-aliased masks) and run --verify_dataset_integrity. nnU-Net expects labels to be consecutive integers starting at 0 and will check for unexpected label values. Also check how many cases in train/val actually contain foreground, because for very sparse vessel tasks the model can otherwise collapse to predicting only background. Pseudo Dice is only computed on validation batches, not on the full validation set, so it can be noisy, but it should usually not remain exactly 0 that long if the dataset is set up correctly.

Best,
Lars
